### ISSUE TITLE
Does CuDNN4 officially support batch normalization?

### ISSUE THREAD

**Issue by fayeshine** [Theano/Theano#3827]
It says on the NVIDIA official site:
"The new cuDNN 4 release delivers faster neural network training and is optimized for Maxwell GPUs.
Train neural networks up to 14x faster using Google’s Batch Normalization technique."

Does it means that NVIDIA implement a high efficiency  batch normalization in cudnn.so, so we only need to call it and don't need to implement by ourselves? Anyone tested the benchmark of NVIDIA's batch normalization and Theano's batch normalization on speed?

**Comment by nouiz** [comment]
cudnn support batch normallization, but we only wrapped cudnn v4
convolution change, not those operation.

It was planned, but we missed time. Maybe we will do it at some point, but
we have other priorities for now.

I suppose it is faster then doing the same with a graph of Theano op, but I
have no benchmark, so I don't know how much faster it would be.

If someone is interrested to work on this, let us know.

**Comment by happygds** [comment]
Could I use cudnn v4 to substitute for cudnn v3 in theano?

**Comment by fayeshine** [comment]
@happygds ,sure, but you need a bleeding edge version of theano

**Comment by happygds** [comment]
@fayeshine hi, I used to extract cudnn v3 to the filefolder `/usr/local/cuda` as the website said. However, when I extracted v4 to replace v3, theano cannot detect cudnn. I tested using this:
> from theano.sandbox.cuda import dnn
> dnn.dnn_available()

It got False. How to install v4 correctly?

**Comment by fayeshine** [comment]
@happygds did you use the newest master branch of theano?if so, you need a a soft link libcudnn.so -> libcudnn.so.4

**Comment by happygds** [comment]
@fayeshine yeah, after deleting the original files and set up the soft link, it worked. Thanks a lot !

**Comment by avostryakov** [comment]
@nouiz I'm interesting to implement it in theano. 

Is theano.sandbox.cuda.dnn right place to implement it? I'll follow GpuDnnSoftmax(GpuDnnSoftmaxBase) and GpuDnnSoftmaxGrad(GpuDnnSoftmaxBase) implementations as an example.

**Comment by avostryakov** [comment]
I create a pull request with wrappers of cuDNN v.4 Batch Normalization functions:
https://github.com/Theano/Theano/pull/4237

It's not ready yet.

**Comment by nouiz** [comment]
it is merged a long time ago.
